File: backend/app/schemas/media_asset.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.enums import EntityType, PhotoApprovalStatus, StorageProvider


class MediaAssetRead(BaseModel):
    """Schema for reading media asset metadata."""

    id: uuid.UUID
    entity_type: EntityType
    entity_id: uuid.UUID | None = None
    storage_provider: StorageProvider
    file_path: str
    url: str
    thumbnail_url: str | None = None
    file_size_bytes: int
    mime_type: str
    width: int | None = None
    height: int | None = None
    status: PhotoApprovalStatus
    rejection_reason: str | None = None
    uploaded_by_id: uuid.UUID
    uploaded_by_name: str | None = None
    approved_by_id: uuid.UUID | None = None
    approved_by_name: str | None = None
    created_at: datetime
    updated_at: datetime
    approved_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class PhotoRejectRequest(BaseModel):
    """Request schema for rejecting a photo asset."""

    rejection_reason: str


class MediaAssetListResponse(BaseModel):
    """Response schema for paginated/filtered media asset lists."""

    items: list[MediaAssetRead]
    total: int
    pending_count: int

File: frontend/src/types/media_asset.ts
export type EntityType =
  | 'RESIDENT'
  | 'VISITOR'
  | 'EMPLOYEE'
  | 'LOT'
  | 'ANNOUNCEMENT'
  | 'OCCURRENCE';

export type StorageProvider =
  | 'LOCAL_DISK'
  | 'VERCEL_BLOB'
  | 'AWS_S3'
  | 'CLOUDINARY';

export type PhotoApprovalStatus =
  | 'PENDING_APPROVAL'
  | 'APPROVED'
  | 'REJECTED';

export interface MediaAssetRead {
  id: string;
  entity_type: EntityType;
  entity_id: string | null;
  storage_provider: StorageProvider;
  file_path: string;
  url: str;
  thumbnail_url: string | null;
  file_size_bytes: number;
  mime_type: string;
  width: number | null;
  height: number | null;
  status: PhotoApprovalStatus;
  rejection_reason: string | null;
  uploaded_by_id: string;
  uploaded_by_name?: string | null;
  approved_by_id?: string | null;
  approved_by_name?: string | null;
  created_at: string;
  updated_at: string;
  approved_at?: string | null;
}

export interface PhotoRejectRequest {
  rejection_reason: string;
}

export interface MediaAssetListResponse {
  items: MediaAssetRead[];
  total: number;
  pending_count: number;
}
